### PR TITLE
Extend the HTTP Probe To Support Different Request Methods, Bodies, and Headers

### LINKED ISSUE #129: Extend HTTP probe type to support more complex probes
It would be useful if the HTTP probe was extended to supported the following:

- All HTTP method types (`GET`, `POST`, `PUT`, `HEAD`, `DELETE`, `PATCH`, `OPTIONS`).
- Request bodies.
- Arbitrary headers.

Being able to customize HTTP requests in these ways would allow for a more comprehensive probe into a system in line with normal user behavior.

### PR DESCRIPTION
**Pull request by MrAlias**
Initial attempt at extending the http probe to support:

- All HTTP method types (GET, POST, PUT, HEAD, DELETE, PATCH, OPTIONS).
- Request bodies.
- Arbitrary headers.

### Adds

- `method`, `headers`, and `body` to the http protobuf definition.
- Unit-tests for new configuration values.

### Updates

- The protoc generated golang code for the http probe proto package.

### Changes

- The http probe to support configuration of request method, headers, and body.
- The http `Probe.Init` function to return errors for configuration errors (in addition to the already logged message).
- The http probe unittests structure to support quick definition of different probe configuration scenarios.

### Related Issues

Resolves #129

### CODE REVIEW

**Comment by manugarg** [comment]
Thanks @MrAlias. I'll take a look at it today.

File: probes/http/proto/config.proto [hunk 1]
```diff
@@ -8,6 +8,21 @@ message ProbeConf {
     HTTPS = 1;
   }
 
+  enum MethodType {
```
Comment (manugarg) [inline probes/http/proto/config.proto]: Just call it Method?

enum Method {
  ..
}

File: probes/http/http.go [hunk 2]
```diff
@@ -120,6 +122,27 @@ func (p *Probe) Init(name string, opts *options.Options) error {
 		p.protocol = "https"
 	default:
 		p.l.Errorf("Invalid Protocol: %s", p.c.GetProtocol())
+		return fmt.Errorf("Invalid Protocol: %s", p.c.GetProtocol())
+	}
+
+	switch p.c.GetMethod() {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply probes/http/http.go]: @manugarg thanks for the suggestion, I've incorporated this change.

I originally copied this from the code above where `p.protocol` is set from a similar `ENUM`.  

What are your thoughts on including a similar change for that line as well in this PR? E.g.

```
--- a/probes/http/http.go
+++ b/probes/http/http.go
@@ -115,16 +115,7 @@ func (p *Probe) Init(name string, opts *options.Options) error {
 
        p.targets = p.opts.Targets.List()
 
-       switch p.c.GetProtocol() {
-       case configpb.ProbeConf_HTTP:
-               p.protocol = "http"
-       case configpb.ProbeConf_HTTPS:
-               p.protocol = "https"
-       default:
-               p.l.Errorf("Invalid Protocol: %s", p.c.GetProtocol())
-               return fmt.Errorf("Invalid Protocol: %s", p.c.GetProtocol())
-       }
-
+       p.protocol = p.c.GetProtocol().String()
        p.method = p.c.GetMethod().String()
        p.url = p.c.GetRelativeUrl()
        if len(p.url) > 0 && p.url[0] != '/' {
```

Or, would that be a bit too far out of scope for this PR?

**Comment by manugarg** [comment]
I am in the process of importing this change internally. I'll update here once it's complete.

**Comment by manugarg** [comment]
Committed the change:
https://github.com/google/cloudprober/commit/2ac8624950cb92fbdc7cf7146acbe0b81d2caa77

Closing this PR.

**Pull request closed without merging**